File: lib/wip/dotenv_loader.rb
# frozen_string_literal: true

require 'pathname'

module Wip
  # Parses a .env file the way `docker compose` does, so values don't have to
  # be duplicated into wip.yml just to reach the container as -e flags.
  class DotenvLoader
    LINE_PATTERN = /\A(?:export\s+)?([A-Za-z_][A-Za-z0-9_]*)=(.*)\z/

    def initialize(path)
      @path = Pathname(path)
    end

    def load
      return {} unless @path.file?

      @path.readlines.each_with_object({}) do |line, env|
        line = line.strip
        next if line.empty? || line.start_with?('#')

        match = LINE_PATTERN.match(line)
        next unless match

        env[match[1]] = unquote(match[2])
      end
    end

    private

    def unquote(value)
      value = value.strip
      return value[1..-2] if value.start_with?('"') && value.end_with?('"') && value.length >= 2
      return value[1..-2] if value.start_with?("'") && value.end_with?("'") && value.length >= 2

      value.split(/\s+#/, 2).first.to_s.strip
    end
  end
end

File: src/Wip.Core/Configuration/DotenvLoader.cs
using System.Text.RegularExpressions;

namespace Wip.Configuration;

/// <summary>
/// Parses a .env file the way <c>docker compose</c> does, so values don't have to be
/// duplicated into wip.yml just to reach the container as <c>-e</c> flags.
/// </summary>
public sealed partial class DotenvLoader
{
    private readonly string path;

    public DotenvLoader(string path) => this.path = path;

    /// <summary>
    /// Returns the parsed pairs in file order. Order is load-bearing: it decides the
    /// sequence of <c>-e</c> flags on the generated wslc command line.
    /// </summary>
    public OrderedDictionary<string, string> Load()
    {
        var result = new OrderedDictionary<string, string>(StringComparer.Ordinal);
        if (!File.Exists(path))
        {
            return result;
        }

        foreach (var rawLine in File.ReadLines(path))
        {
            var line = rawLine.Trim();
            if (line.Length == 0 || line.StartsWith('#'))
            {
                continue;
            }

            var match = LinePattern().Match(line);
            if (!match.Success)
            {
                continue;
            }

            result[match.Groups[1].Value] = Unquote(match.Groups[2].Value);
        }

        return result;
    }

    private static string Unquote(string raw)
    {
        var value = raw.Trim();
        if (value.Length >= 2 &&
            ((value.StartsWith('"') && value.EndsWith('"')) || (value.StartsWith('\'') && value.EndsWith('\''))))
        {
            return value[1..^1];
        }

        // An unquoted value ends at the first whitespace-preceded '#'. A '#' with no
        // space before it is part of the value, matching compose's own reading.
        var comment = TrailingComment().Match(value);
        return (comment.Success ? value[..comment.Index] : value).Trim();
    }

    [GeneratedRegex(@"\A(?:export\s+)?([A-Za-z_][A-Za-z0-9_]*)=(.*)\z")]
    private static partial Regex LinePattern();

    [GeneratedRegex(@"\s+#")]
    private static partial Regex TrailingComment();
}
